### DISCUSSION TITLE
How to install "flet pack" command for packaging desktop app into an executable?

### DISCUSSION THREAD

**Discussion by TheCodeCache** [flet-dev/flet#979] [General]
The [packaging-desktop-app](https://flet.dev/docs/guides/python/packaging-desktop-app) docs talk about packaging a desktop app into an executable using `flet pack` command as follows - 

`flet pack your_program.py`

My **query** : how do we get this `flet pack` command installed on Windows, what steps need to be followed ?

Please help to resolve it, Thanks!

**Comment by iron3oxide** [comment]
Normally the command should automatically be available when you run `pip install flet`. If your shell (Powershell I presume) doesn't find it, check that you are in the environment where you installed flet (you might have installed it in a virtual environment that is not active or outside one you are currently in). If that checks out, try running `py -m flet pack your_program.py`, that might also help.

**Reply by TheCodeCache** [reply]
I'm not using virtual env per se, rather using system env for python interpreter

so, I tried running the below cmd in PowerShell but getting below error - 

PS C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test> py -m flet pack .\hello_flet.py
C:\Program Files\Python310\python.exe: No module named flet.__main__; 'flet' is a package and cannot be directly executed

please suggest

**Reply by iron3oxide** [reply]
Can you show me the output of `pip show flet` ? There have recently been changes to the project structure and this might have broken the command, but it is important to establish the version of flet you're running first.

**Reply by TheCodeCache** [reply]
PFB - 

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>pip show flet
Name: flet
Version: 0.3.2
Summary: Flet for Python - easily build interactive multi-platform apps in Python
Home-page:
Author:
Author-email: Appveyor Systems Inc. <hello@flet.dev>
License: MIT
Location: c:\users\manor\appdata\roaming\python\python310\site-packages
Requires: beartype, oauthlib, packaging, repath, requests, watchdog, websocket-client
Required-by:

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>

**Reply by iron3oxide** [reply]
I see. Now run `pip install --upgrade --pre flet`. This should upgrade flet to the most recent (prerelease) version. Then try the flet pack command again and tell me if it works now.

**Reply by DavMelchi** [reply]
https://youtu.be/PTqCtWhWgbk

Have a look on this video and setup your path accordingly.

**Reply by TheCodeCache** [reply]
> I see. Now run `pip install --upgrade --pre flet`. This should upgrade flet to the most recent (prerelease) version. Then try the flet pack command again and tell me if it works now.

it still fails unfortunately - 

PFB the logs - 

C:\Users\manor>cd C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>py -m flet pack hello_flet.py
C:\Program Files\Python310\python.exe: No module named flet.__main__; 'flet' is a package and cannot be directly executed

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>pip show flet
Name: flet
Version: 0.4.0.dev1151
Summary: Flet for Python - easily build interactive multi-platform apps in Python
Home-page:
Author: Appveyor Systems Inc.
Author-email: hello@flet.dev
License: MIT
Location: c:\users\manor\appdata\roaming\python\python310\site-packages
Requires: flet-core, httpx, oauthlib, packaging, watchdog, websocket-client, websockets
Required-by:

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>

**Reply by TheCodeCache** [reply]
> https://youtu.be/PTqCtWhWgbk
> 
> Have a look on this video and setup your path accordingly.

I tried this link to set up the path in sys env variable, yet it failed to work, it still does not recognize the `flet` command

PFB the error logs - 

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>flet pack hello_flet.py
'flet' is not recognized as an internal or external command,
operable program or batch file.

PFB the screenshot from Environment variable - 
![image](https://user-images.githubusercontent.com/26399543/216792138-76260e33-be22-4f03-8d4f-44b2a02854f8.png)

**Reply by FeodorFitsner** [reply]
What do you have when running `where python`?

**Reply by TheCodeCache** [reply]
> What do you have when running `where python`?

PFB - 

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>where python
C:\Program Files\Python310\python.exe
C:\Users\manor\AppData\Local\Microsoft\WindowsApps\python.exe

**Reply by FeodorFitsner** [reply]
so what do you have in `dir c:\users\manor\appdata\roaming\python\python310`?

**Reply by TheCodeCache** [reply]
> so what do you have in `dir c:\users\manor\appdata\roaming\python\python310`?

These are the only 3 folders I have - 

![image](https://user-images.githubusercontent.com/26399543/216792303-ab3415cc-9c2f-482c-8bb9-c690393f107c.png)

However, I do see some of the `flet` package folders if I navigate inside site-packages - 
PFB
![image](https://user-images.githubusercontent.com/26399543/216792363-7985961a-6f68-421f-8372-053706977d52.png)

**Reply by FeodorFitsner** [reply]
Alright, and finally if you do `dir c:\users\manor\appdata\roaming\python\python310\Scripts` I guess you'll see `flet.exe` there.

If so, you need to add `c:\users\manor\appdata\roaming\python\python310\Scripts` to `PATH` environment variable.

**Reply by TheCodeCache** [reply]
> Alright, and finally if you do `dir c:\users\manor\appdata\roaming\python\python310\Scripts` I guess you'll see `flet.exe` there.
> 
> If so, you need to add `c:\users\manor\appdata\roaming\python\python310\Scripts` to `PATH` environment variable.

Thank you so much @FeodorFitsner 
I'm able to locate the flet.exe, updated the `PATH` env variable and tried creating the executable from code, but it fails on `controls` module not found, 

This is the folder structure I've in my local PyCharm IDE - 

![image](https://user-images.githubusercontent.com/26399543/216792910-9b4468b1-5ee5-4237-aff2-8f08bcb46770.png)

When I run `flet pack hello_flet.py` from the directory containing hello_flet.py, I get below error - 

C:\Users\manor\OneDrive\Desktop\dev\code\python_apps\com\test>flet pack hello_flet.py

![image](https://user-images.githubusercontent.com/26399543/216793003-37425396-cc10-48bf-a20e-1ba58fa326d2.png)

Please let me know if you need command logs to get some hint into this issue, thank you!

**Reply by iron3oxide** [reply]
That's what happens when you use relative imports. `hello_flet.py` will search in the right path, but `pyinstaller` (which flet pack uses under the hood) will look for a `controls` module in its own directory. I always try to use absolute imports for this same reason. 

1. Make the test package installable by adding setup files next to it in the com directory ([here's a guide](https://godatadriven.com/blog/a-practical-guide-to-setuptools-and-pyproject-toml/))
2. Run `pip install -e .` in the com directory. This will install the test package in editable mode. Caution: this will install it systemwide if you are not in a virtual environment.
3. Instead of `from controls.xyz import abc`, write `from test.controls.xyz import abc`

**Reply by TheCodeCache** [reply]
Thanks [iron3oxide](https://github.com/iron3oxide), will try this out and get back in case of any issues
